---
title: 'Chapter 5: 16S data analysis using the R package Phyloseq'
author: "Emily Giroux"
date: "4/17/2019"
output: pdf_document
fontsize: 11pt
geometry: margin=1in
urlcolor: blue
header-includes: \usepackage{xcolor}
---

```{r, global_options, eval=TRUE, echo=FALSE, cache=TRUE}
#Set the global options for knitr
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff=80), tidy = TRUE, fig.align='center',
               cache=FALSE, collapse=TRUE, echo=FALSE, eval=FALSE, include=FALSE,
               message=FALSE, quietly=TRUE, results='hide', warn.conflicts=FALSE, 
               warning=FALSE)
```

**Using package `BiocManager` to install required packages:**
```{r, biocInstall, eval=TRUE, echo=TRUE, include=TRUE, cache=TRUE, tidy=FALSE, message=FALSE}
r <- getOption("repos")
r["CRAN"] <- "http://cran.us.r-project.org"
options(repos = r)

if (!requireNamespace("BiocManager"))
    install.packages("BiocManager")
BiocManager::install()

library("BiocManager")
.cran_packages <- c("cowplot", "data.table", "ggplot2", "knitr", "rprojroot")
.bioc_packages <- c("BiocStyle", "Biostrings", "dada2", 
                    "ShortRead")
.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
   install.packages(.cran_packages[!.inst])
}
.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
  BiocManager::install(.bioc_packages[!.inst], ask = FALSE)
}
```

**Load packages into session, and print package versions:**
```{r, showBiocPackages, echo=TRUE, eval=TRUE, include=TRUE, results='hold', cache=TRUE}
sapply(c(.cran_packages, .bioc_packages), require, character.only = TRUE)
```

```{r sourcing_my_functions, echo=FALSE, eval=TRUE, include=FALSE, cache=TRUE}
#Source our custom R scripts:    
#For this we will use the rprojroot package to set the directory structures. This will help us when finding our files to source functions. We specify ours is an RStudio project. The root object contains a function that will help us locate our package R files regarless of our current working directory.
library("rprojroot")
root <- rprojroot::is_rstudio_project
scriptsPath <- root$make_fix_file(".")("R")
scripts  <- dir(root$find_file("R", path = root$find_file()))
scriptsl <- paste(scriptsPath, scripts, sep = "//")
lapply(scriptsl, source)
# Record the path to the environment images directory:
sharedPath <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/PIRL_working_directory/"
analysis <- "ecobiomics/"
sharedPathAn <- paste(sharedPath, analysis, sep = "")
imageDirPath <- "/home/CFIA-ACIA/girouxeml/GitHub_Repos/r_environments/ecobiomics/"

# Recall from Chapter 4b 16S DADA2-specific processing step: The only variables we need from the previous environment for this analysis chapter are the "ps" variable from the fitGTR image, which we re-saved in the chptImageB variable:
chptImageB <- "ecobiomics_16S_2b.RData"
load(paste(imageDirPath, chptImageB, sep = ""))

# Fresh image for this new chapter, and image to continue saving going forward:
chptImage    <- "ecobiomics_16S_analysis.RData"
save.image(paste(imageDirPath, chptImage, sep = ""))
```
\pagebreak    

When re-loading this chapter to continue an already started analysis:
```{r}
sharedPath <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/PIRL_working_directory/"
analysis <- "ecobiomics/"
sharedPathAn <- paste(sharedPath, analysis, sep = "")
imageDirPath <- "/home/CFIA-ACIA/girouxeml/GitHub_Repos/r_environments/ecobiomics/"
chptImage    <- "ecobiomics_16S_analysis.RData"
load(paste(imageDirPath, chptImage, sep = ""))
```

Read in both sample metadata tables generated at the end of read processing:
```{r, echo=TRUE, eval=FALSE, message=FALSE, warning=FALSE, tidy=FALSE}
library(data.table)
sampleTblName <- "final_processed_metadata"
metadata      <- fread(paste(sharedPathReg, sampleTblName, ".csv", sep = ""), 
                       sep = "auto", header = TRUE)
sampleData <- read.table(paste(sharedPathReg, sampleTblName, "_forPhyloseq.txt", sep = ""))
```

Let's get familiar with our phyloseq object created at the end of our sequencing sample processing chapter:     
Below I am using the ps, rather than phySeq objects. Recall from the last chunks of Chapter 2, the phySeq object is a phyloseq object without the fitGTR$tree info, while the ps object was created leveraging the fitGTR$tree information. The phySeq object can be used instead of the ps object if the optim.pml command wasn't run.
```{r, testCmds, eval=FALSE, include=FALSE, echo=FALSE, message=FALSE}
library(phyloseq)
rank_names(ps) # Note that for 16S data there are no determinations in the ps variable that extend to the species level
table(tax_table(ps)[, "Genus"], exclude = NULL)
```
Here is if we filter based on having to know the genus:
```{r, testCmds2, eval=FALSE, include=FALSE, echo=-5, message=FALSE, comment=NA}
ps2Genus <- subset_taxa(ps, !is.na(Genus) & !Genus %in% c("", "uncharacterized"))
rank_names(ps2Genus)
table(tax_table(ps2Genus)[, "Genus"], exclude = NULL)
```

Visualize alpha-diversity, genus:
```{r, aDiv1, eval=TRUE, include=TRUE, echo=TRUE, cache=TRUE, message=FALSE}
library(phyloseq)
library(ggplot2)
library(cowplot)
plot_richness(ps2Genus,
              x = "ExtractionKit",
              measures = c("Shannon", "Simpson"),
              color = "Sample") +
                theme(axis.text.x = element_text(angle = 90)) +
  scale_x_discrete(name = "Extraction Kit")
```


Prevalence evaluation for genera:
```{r, prevTblGenera, echo=TRUE, eval=TRUE, cache=TRUE, results='hold', include=TRUE, tidy=FALSE, message=FALSE, comment=NA}
library(phyloseq)
prevDf <- apply(X = otu_table(ps2Genus),
                MARGIN = ifelse(taxa_are_rows(ps2Genus),
                                yes = 1, no = 2),
                FUN = function(x){sum(x>0)})

prevDf <- data.frame(Prevalence = prevDf,
                     TotalAbundance = taxa_sums(ps2Genus),
                     tax_table(ps2Genus))

prevalenceTblGenera <- plyr::ddply(prevDf, "Genus",
                                   function(df1){
                                     cbind(mean(df1$Prevalence),
                                           sum(df1$Prevalence))})
colnames(prevalenceTblGenera) <- c("Genus", "Mean", "Sum")
prevalenceTblGenera
```
### Analysing the 16S data for soil samples only:    
```{r}
psSoil <- subset_samples(ps, Sample=="soil")
```

Generate a table that shows the prevalence of genera present across soil samples:
```{r}
prevDfSoil <- apply(X = otu_table(psSoil),
                    MARGIN = ifelse(taxa_are_rows(psSoil),
                                    yes = 1, no = 2),
                    FUN = function(x){sum(x>0)})

prevDfSoil <- data.frame(Prevalence = prevDfSoil,
                         TotalAbundance = taxa_sums(psSoil),
                         tax_table(psSoil))

prevTblGenSoil <- plyr::ddply(prevDfSoil, "Genus",
                              function(df1){
                                cbind(mean(df1$Prevalence),
                                      sum(df1$Prevalence))})
colnames(prevTblGenSoil) <- c("Genus", "Mean", "Sum")
prevTblGenSoil
```

Filter entries with unidentified Phylum (if you want to analyse only taxa that are resolved down to the genus level, can set that here in place of phylum):
```{r, filterLowPhyla, echo=TRUE, include=TRUE, eval=TRUE, tidy=FALSE, cache=TRUE, results='hold', message=FALSE, comment=NA}
library(phyloseq)
prevTblPhySoil <- plyr::ddply(prevDfSoil, "Phylum",
                              function(df1){
                                cbind(mean(df1$Prevalence),
                                      sum(df1$Prevalence))})
colnames(prevTblPhySoil) <- c("Phylum", "Mean", "Sum")

# See the prevalence of phylum across soil samples:
arrange(prevTblPhySoil, desc(Sum)) 
```

```{r}
# See those phyla that are of low-abundance and appear in less than 10 samples: 
arrange(filter(prevTblPhySoil, Sum < 10))
```
Filter entries with unidentified Phylum, or those phyla that appear in less than 10 samples:
```{r}
library(phyloseq)
phylas <- subset(prevTblPhySoil, prevTblPhySoil$Sum < 10)
psSoilPhy1    <- subset_taxa(psSoil, !Phylum %in% phylas$Phylum)
rank_names(psSoilPhy1)
table(tax_table(psSoilPhy1)[, "Phylum"], exclude = c(NA, NULL))
```
Plot Phylum:
```{r, plotPhylum, echo=TRUE, eval=TRUE, cache=TRUE, tidy=FALSE, results='hold', include=TRUE, message=FALSE}
library(phyloseq)
library(ggplot2)
prevTblPhySoil1 <- subset(prevDfSoil, Phylum %in% get_taxa_unique(psSoilPhy1, "Phylum"))
prevTblPhySoil2 <- subset(prevTblPhySoil1, !is.na(Phylum)) # filtering out taxa where Phylum is NA

ggplot(prevTblPhySoil2, aes(TotalAbundance, Prevalence / nsamples(psSoilPhy1),
                            color = Phylum)) +
  geom_hline(yintercept = 0.05, alpha = 0.5, linetype = 2) +
  geom_point(size = 0.5, alpha = 0.6) +
  scale_x_log10() + xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) + theme(legend.position = "none")
```
Each point in the above plots is a different taxa, Phylum.
```{r, prevTreshSet, eval=TRUE, echo=TRUE, cache=TRUE, comment=NA}
prevalenceThreshold = 0.05*nsamples(psSoilPhy1)
prevalenceThreshold
```
The taxa with a prevalence threshold less than the one set in the above chunk are removed using prune_taxa and put into a new phyloseq object, `ps2`, and we look at the resulting richness plot:    
Note, NucleoSpin was not used for DNA extraction in soil samples.
```{r, taxPrevTresh, echo=TRUE, eval=TRUE, results='hold', cache=TRUE, tidy=FALSE, message=FALSE, comment=NA}
library(phyloseq)
keepTaxa <- rownames(prevTblPhySoil2)[(prevTblPhySoil2$Prevalence >= prevalenceThreshold)]
ps2 <- prune_taxa(keepTaxa, psSoilPhy1)
rank_names(ps2)
table(tax_table(ps2)[, "Phylum"])
plot_richness(ps2, 
              x = "ExtractionKit", 
              measures = c("Shannon", "Simpson"), 
              color = "Sample") +
                theme(axis.text.x = element_text(angle = 90)) +
  scale_x_discrete(name = "Extraction Kit")
```

Phylum - curiosity:     
Curious about the mean and sum prevalence after keeping only taxa passing prevalenceThreshold:
```{r, prevPhylcheck, eval=TRUE, echo=TRUE, cache=TRUE, tidy=FALSE, message=FALSE, comment=NA}
library(phyloseq)
prevDf2 <- apply(X = otu_table(ps2),
                MARGIN = ifelse(taxa_are_rows(ps2), yes = 1, no = 2),
                FUN = function(x){sum(x > 0)})

prevDf2 <- data.frame(Prevalence = prevDf2,
                      TotalAbundance = taxa_sums(ps2),
                      tax_table(ps2))

prevPhylatblThreshold <- plyr::ddply(prevDf2, "Phylum", 
                                     function(df1){
                                       cbind(mean(df1$Prevalence), 
                                             sum(df1$Prevalence))})
colnames(prevPhylatblThreshold) <- c("Phylum", "Mean", "Sum")
prevPhylatblThreshold
```
Note: I am assuming that the mean is the mean number of times the phylum was seen in a sample for all samples in which it was seen, while the sum is the total times it was seen across all samples. For example, if Actinobacteria was seen a total of 179 times, with about 11 occurances per sample, while Armatimonadetes was seen 9 times total and the mean is simply 9 because when it was seen, it was all three in one sample.     
     
Number of unique phyla, genera and species, across all samples:
```{r, taxUniqueRanks, echo=TRUE, eval=TRUE, cache=TRUE, comment=NA}
library(phyloseq)
uniqueClasses <- c("Phylum", "Genus")
for(i in unique(uniqueClasses)) 
  cat(cat(i), length(get_taxa_unique(ps2, taxonomic.rank = i)), "\n")
```

The tax_glom function of phyloseq merges species that have the same taxonomy at certain taxonomic rank, using categorical data. The tip_glom function agglomerates tree tips into a single taxa if they are separated by less than a height specified by `h`. 
```{r, taxGlom, echo=TRUE, eval=TRUE, cache=TRUE}
library(phyloseq)
ps3 <- tax_glom(ps2, "Genus", NArm = TRUE)
h1 = 0.01
ps4 <- tip_glom(ps2, h = h1)
```

How many taxa are left before/after agglomeration?:
```{r}
ntaxa(ps2); ntaxa(ps3); ntaxa(ps4)
```

Out of curiosity how many "reads" does this leave us at???
```{r}
sum(colSums(otu_table(ps2))); sum(colSums(otu_table(ps3))); sum(colSums(otu_table(ps4)))
```

Below we will look at plots of our trees before agglomeration, with agglomeration using tax_glom, and with agglomeration by tip separation using tip_glom:
```{r, tipGlom, echo=TRUE, eval=TRUE, cache=TRUE, tidy=FALSE, message=FALSE}
library(phyloseq)
library(ggplot2)
library(cowplot)
multiPlotTitleTextSize = 15
p2Tree <- plot_tree(ps2, method = "treeonly",
                    ladderize = "left",
                    title = "Before Agglomeration") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))

p3Tree <- plot_tree(ps3, method = "treeonly",
                    ladderize = "left", title = "By Genus") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))

p4Tree <- plot_tree(ps4, method = "treeonly",
                    ladderize = "left", title = "By Height") +
  theme(plot.title = element_text(size = multiPlotTitleTextSize))
gridExtra::grid.arrange(nrow = 1, p2Tree, p3Tree, p4Tree)
```

From here on we will continue using the ps3 with the gglomerated taxa at the Genus level (combine all with the same name) and remove all taxa without genus level assignmentand that has had the `NA', low-abundance, and prevalence threshold filters applied.  
```{r, plotPhylum2, echo=TRUE, eval=TRUE, cache=TRUE, tidy=FALSE, message=FALSE, comment=NA}
library(phyloseq)
library(ggplot2)
plotPhylum <- plot_bar(ps3, x = "ExtractionKit", fill = "Phylum", 
                       facet_grid = ~sample_Sample)
plotPhylum + geom_bar(aes(color = Phylum, fill = Phylum), 
                      stat = "identity", 
                      position = "stack")
```

```{r}
qplot(rowSums(otu_table(ps3)),bins=30) +
  xlab("Logged counts-per-sample")
```

Now lets filter out samples (outliers and low performing samples)
Do some simple ordination looking for outlier samples, first we variance stabilize the data with a log transform, then perform PCoA using bray’s distances

```{r}
logt  = transform_sample_counts(ps3, function(x) log(1 + x) )
out.pcoa.logt <- ordinate(logt, method = "PCoA", distance = "bray")
evals <- out.pcoa.logt$values$Eigenvalues
plot_ordination(logt, out.pcoa.logt, type = "Sample", 
                color = "ExtractionKit") + labs(col = "ExtractionKit") +
  coord_fixed(sqrt(evals[2] / evals[1]))
```
  
```{r}
plot_ordination(logt, out.pcoa.logt, type = "taxa", color = "Phylum") 
```
```{r}
coord_fixed(sqrt(evals[2] / evals[1]))
```
Look for low performing samples
```{r}
qplot(rowSums(otu_table(ps3)),bins=30) +
  xlab("Logged counts-per-sample")
```
Show taxa proportions per sample

```{r}
grid.arrange(nrow = 2,
qplot(as(otu_table(logt),"matrix")["16S_NA_GB_K1A_2_FD",], geom = "histogram", bins=30) +
  xlab("Relative abundance"),

qplot(as(otu_table(logt),"matrix")["16S_NA_GB_K1B_2_FD",], geom = "histogram", bins=30) +
  xlab("Relative abundance")
)

```
##### NOTE   - may see this as the start of a problem when trying to follow the tutorial at "https://ucdavis-bioinformatics-training.github.io/2017-September-Microbial-Community-Analysis-Workshop/friday/MCA_Workshop_R/phyloseq.html"    
There appears to be a difference where their taxa are as rows and mine are as columns???    
So in the chunk below, keeping taxa seems to save samples instead of taxa, when the tutorial seems to have meant the reverse... This error propagates more uncertainty and issues running commands as we progress from here on in.

```{r}
prop  = transform_sample_counts(ps3, function(x) x / sum(x) )
keepTaxa <- ((apply(otu_table(prop) >= 0.005,1,sum,na.rm=TRUE) > 2) | (apply(otu_table(prop) >= 0.05, 1, sum,na.rm=TRUE) > 0))
table(keepTaxa)
```

The following works when viewing by kit, but is too messy when viewing by sample since we have 53 samples, and after pruning out low log counts samples, still have 29 samples remaining, which makes the x-axis useless to try to visualise.
```{r}
## for Firmictures
plot_abundance = function(physeq, meta, title = "",
                 Facet = "Order", Color = "Order"){
  # Arbitrary subset, based on Phylum, for plotting
  p1f = subset_taxa(physeq, Phylum %in% c("Firmicutes")) # I changed "p__Firmicutes" to just "Firmicutes" or else command fails with dimnames as this table is more processee than the example one in the tutorial
  mphyseq = psmelt(p1f)
  mphyseq <- subset(mphyseq, Abundance > 0)
  ggplot(data = mphyseq, mapping = aes_string(x = meta,y = "Abundance",
                                 color = Color, fill = Color)) +
    geom_violin(fill = NA) +
    geom_point(size = 1, alpha = 0.3,
                position = position_jitter(width = 0.3)) +
    facet_wrap(facets = Facet) + scale_y_log10()+
    theme(legend.position="none")
}

# remove samples with fewer than 10000 logged counts per sample
ps3.4 <- prune_samples(sample_sums(ps3)>=10000, ps3) 
ps3.4

# transform counts into "abundances"
ps3.4ra <- transform_sample_counts(ps3.4, function(x){x / sum(x)})
ps3.4hell <- ps3.4ra
otu_table(ps3.4hell) <- otu_table(decostand(otu_table(ps3.4hell), method = "hellinger"), taxa_are_rows=FALSE)
ps3.4log <- transform_sample_counts(ps3.4, function(x) log(1 + x))

plotOriginal = plot_abundance(ps3.4, "ExtractionKit", title="original") # Viewing per extraction kit easier than per sample, but we know that the FastDNA has the lowest number of samples, and even loses some after pruning.
plotRelative = plot_abundance(ps3.4ra, "ExtractionKit", title="relative")
plotHellinger = plot_abundance(ps3.4hell, "ExtractionKit", title="Hellinger")
plotLog = plot_abundance(ps3.4log, "ExtractionKit", title="Log")
# Combine each plot into one graphic.
grid.arrange(nrow = 4, plotOriginal, plotRelative, plotHellinger, plotLog)
```
One at a time:
```{r}
grid.arrange(nrow = 1, plotOriginal)
```

```{r}
grid.arrange(nrow = 1, plotRelative)
```

```{r}
grid.arrange(nrow = 1, plotHellinger)
```

```{r}
grid.arrange(nrow = 1, plotLog)
```


```{r}
ps3.4hell_trim <- prune_taxa(keepTaxa, ps3.4hell)
plot_heatmap(ps3.4hell_trim, "PCoA", distance="bray", sample.label="SampleName", taxa.label="Genus", low="#FFFFCC", high="#000033", na.value="white")
```



```{r, topTaxPlotsSpp, echo=TRUE, eval=TRUE, cache=TRUE, tidy=FALSE, message=FALSE, comment=NA}
library(phyloseq)
library(ggplot2)

topSpecies <- names(sort(taxa_sums(ps2), TRUE)[1:27])
taxTabSpp  <- cbind(tax_table(ps2), Species = NA)
taxTabSpp[topSpecies, "Species"] <- as(tax_table(ps2)[topSpecies, "Species"],
                                             "character")
tax_table(ps2) <- tax_table(taxTabSpp)
ps2mSpp <- merge_samples(ps2, "ExtractionKit")
sample_data(ps2mSpp)$ExtractionKit <- levels(sample_data(ps2)$ExtractionKit)
ps2mSpp <- transform_sample_counts(ps2mSpp, function(x) 100 * x/sum(x))

ps2mSppTop = prune_taxa(topSpecies, ps2mSpp)
title = "Figure 2 Top 17 Species, attempt 1"
plotSpecies <- plot_bar(ps2mSppTop, "ExtractionKit", 
                        fill = "Species", 
                        title = title) + 
  coord_flip() + 
  ylab("Percentage of Sequences") + ylim(0, 70) +
  geom_bar(aes(color = Species, fill = Species), 
           stat = "identity", position = "stack")
plotSpecies
```
Below I'm testing what a heatmap would look like for taxa abundance across extraction kits:
```{r, plotHeatMap, echo=TRUE, eval=TRUE, cache=TRUE, tidy=FALSE, message=FALSE, comment=NA, warning=FALSE}
library(phyloseq)
library(ggplot2)
plot_heatmap(ps2mTop, "PCoA", distance="bray", 
             sample.label="ExtractionKit", 
             taxa.label="Species",
             low="#66CCFF", high="#000033", na.value="white") +
  scale_x_discrete(expand=c(0,0))
```

```{r, barPlotTop20Genus, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, tidy=FALSE, cache=TRUE}
library(phyloseq)
library(ggplot2)

top20 <- names(sort(taxa_sums(ps2), decreasing=TRUE))[1:20]
ps.top20 <- transform_sample_counts(ps2, function(OTU) OTU/sum(OTU))
ps.top20 <- prune_taxa(top20, ps.top20)

plotGenus <- plot_bar(ps.top20, x = "ExtractionKit", fill = "Genus", 
                      facet_grid = ~sample_Sample)
plotGenus + geom_bar(aes(color = Genus, fill = Genus), 
                     stat = "identity", 
                     position = "stack")
```

Not yet working:     
Transform data to proportions as appropriate for Bray-Curtis distances:
```{r, trnfrmBrayC, eval=FALSE, echo=TRUE, tidy=FALSE, warning=FALSE, error=FALSE}
library(phyloseq)
psProp <- transformSampleCounts(ps2, function(otu) otu/sum(otu))

ordNmdsBray1 <- ordinate(psProp, method="NMDS", distance="bray")
```

Statistical analysis
Post hoc comparisons between the four tested methods were made using the Tukey HSD test. OTUs or OTUs pooled at phylum, class, order, family or genera level with different abundances were identified using a generalized linear model where the counts follow an overdispersed Poisson distribution (Kristiansson, Hugenholtz and Dalevi 2009; Jonsson et al.2016). The p-values were corrected for multiple testing using the false discovery rate (FDR) method. The OTU abundance was used for principal component analysis (PCA). Shared OTUs between DNA extraction methods were graphically visualised in Venn diagrams using the corresponding OTU tables exported from QIIME. The hypergeometric distribution was used to test the distribution of gram negatives and gram positives among the taxa identified with the respective four DNA extraction methods. Pearson correlations were used to test for correlations between descriptors of DNA quantity and quality (Table 1), and descriptors of taxonomic diversity (Table 2). The statistical significance for all the analyses was set to P < 0.05 or FDR< 0.05. All statistical analyses were carried out using the R v.3.2.0 software (R Core Team 2013).

Table 2.
Detected 16S rRNA richness and biodiversity from marine periphyton biofilm DNA extracted with the four studied methods.

 	FastDNA Soil 	PowerPlant 	PowerBiofilm 	PlantDNAzol 	 
n 	3 	2 	3 	3 	P-values 
No. of OTUs 	666 ± 42 	704 ± 58 	809 ± 11 	791 ± 7 	P < 0.05 
No. of phyla 	17 ± 1 	17 ± 1 	17 ± 1 	18 ± 0 	ns 
No. of classes 	39 ± 1 	40 ± 1 	40 ± 2 	41 ± 1 	ns 
No. of orders 	68 ± 3 	70 ± 3 	71 ± 2 	72 ± 1 	ns 
No. of families 	91 ± 2 	95 ± 6 	104 ± 3 	106 ± 2 	P < 0.05 
No. of genera 	141 ± 4 	145 ± 12 	159 ± 4 	162 ± 3 	P < 0.05 

Each value represents the arithmetic mean ± standard error of the mean. n: number of replicates. Statistical significance between extraction methods is denoted as P < 0.05 (ANOVA).

ns: indicates no statistically significant differences between extraction methods.

