---
title: "OakWilt"
author: "Emily Giroux"
date: "9/30/2020"
output:
  pdf_document: default
  html_document: default
urlcolor: blue
header-includes: \usepackage{xcolor}
---

```{r, global_options, eval=TRUE, echo=FALSE}
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff = 80), tidy = TRUE, fig.align = 'center',
               cache = FALSE, collapse = TRUE, echo = FALSE, eval = FALSE, include = FALSE,
               message = FALSE, quietly = TRUE, results = 'hide', warn.conflicts = FALSE, 
               warning = FALSE)
```

**Using package `BiocManager` to install required packages:**
```{r, biocInstall, eval=TRUE, echo=TRUE, include=TRUE}
#Installing required packages
r <- getOption("repos")
r["CRAN"] <- "http://cran.us.r-project.org"
options(repos = r)

if (!requireNamespace("BiocManager"))
    install.packages("BiocManager")
BiocManager::install()

library("BiocManager")
.cran_packages <- c("data.table", "kableExtra", "knitr", "rprojroot")
.bioc_packages <- c("BiocStyle", "Biostrings", "dada2", "RAM")
.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
   install.packages(.cran_packages[!.inst])
}
.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
  BiocManager::install(.bioc_packages[!.inst], ask = FALSE)
}
```
   
**Load packages into session, and print package versions:**
```{r, showBiocPackages, echo=TRUE, eval=TRUE, include=TRUE, results='hold'}
sapply(c(.cran_packages, .bioc_packages), require, character.only = TRUE)
```
**Source our custom R scripts:**    
For this we will use the rprojroot package to set the directory structures. This will help us when finding our files to source functions. We specify ours is an RStudio project. The root object contains a function that will help us locate our package R files regarless of our current working directory.
```{r sourcing_my_functions, echo=TRUE, eval=TRUE, include=TRUE, tidy=FALSE}
library("rprojroot")
root        <- rprojroot::is_rstudio_project
scriptsPath <- root$make_fix_file(".")("R")
scripts     <- dir(root$find_file("R", path = root$find_file()))
scriptsl    <- paste(scriptsPath, scripts, sep = "/")
lapply(scriptsl, source)
```

Setting up working directories:
```{r}
sharedPath <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/PIRL_working_directory"
analysis <- "oakwilt"
sharedPathAn <- paste(sharedPath, analysis, sep = "/")
dir.create(sharedPathAn, showWarnings = TRUE, recursive = FALSE)
imageDirPath <- "/home/CFIA-ACIA/girouxeml/GitHub_Repos/r_environments/oakwilt/"
dir.create("/home/CFIA-ACIA/girouxeml/GitHub_Repos/r_environments/oakwilt", 
           showWarnings = TRUE, recursive = FALSE)
baseImage <- "oakwilt.RData"
save.image(paste(imageDirPath, baseImage, sep = ""))
```
Quick image load:
```{r}
imageDirPath <- "/home/CFIA-ACIA/girouxeml/GitHub_Repos/r_environments/oakwilt/"
baseImage <- "oakwilt.RData"
load(paste(imageDirPath, baseImage, sep = ""))
```
### Step 1:       
Set up all folders (baseDir, qiime2, trimmed, logs)     
```{r}
library("data.table")
rawDataDir <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/data/forMarco"
compressedFiles <- list.files(rawDataDir, pattern = "*.bz2", full.names = TRUE)
metadata <- as.data.table(cbind(compressedFiles))
metadata$rawFileName <- basename(metadata$compressedFiles)
metadata$basename <- gsub(".tar.bz2", "", metadata$rawFileName)
rawDataWorkingPath <- paste(sharedPathAn, "rawData", sep = "/")
dir.create(rawDataWorkingPath, showWarnings = TRUE, recursive = FALSE)
metadata$rawWorkingPath <- paste(rawDataWorkingPath, metadata$basename, sep = "/")

for(i in 1:nrow(metadata)){
  cmd[i] <- paste("mkdir -p ",  rawDataWorkingPath, " && tar -xvjf ", metadata$compressedFiles[i], 
                  " -C ", rawDataWorkingPath, sep = "")  
  system(cmd[i])
}

metadataITSF_files <- list.files(rawDataWorkingPath, pattern = "ITSF", recursive = TRUE, full.names = TRUE)
metadataITSF <- as.data.table(cbind(metadataITSF_files))
metadataITSF$basename <- basename(metadataITSF$metadataITSF_files)
metadataITSF$barcode <- gsub(".*ITS1F_", "", metadataITSF$basename)
metadataITSF$barcode <- gsub(".fastq", "", metadataITSF$barcode)
metadataITSF$barcode <- gsub("b", "B", metadataITSF$barcode)

metadataITS2R_files <- list.files(rawDataWorkingPath, pattern = "ITS2rev", recursive = TRUE, full.names = TRUE)
metadataITS2R <- as.data.table(cbind(metadataITS2R_files))
metadataITS2R$basename <- basename(metadataITS2R$metadataITS2R_files)
metadataITS2R$barcode <- gsub(".*ITS2_A_", "", metadataITS2R$basename)
metadataITS2R$barcode <- gsub(".fastq", "", metadataITS2R$barcode)

# Join the metadata samples from the forward and reverse tables using the common barcode to join rows:
setkey(metadataITSF, barcode)
setkey(metadataITS2R, barcode)

metadataITS <- merge(metadataITSF, metadataITS2R, all.x = TRUE)
setnames(metadataITS, "basename.x", "fwdFastq")
setnames(metadataITS, "basename.y", "revFastq")
metadataITS <- na.omit(metadataITS)
```

Prepare file directories:    
Qiime2     
# Input folder     
export fastq=/media/30tb_raid10/data/PIRL/2020-01-15_OAK_ITSF_30     
# Output folder     
export baseDir=/media/2TB_NVMe/pirl_2020-01-15_ITS1F     
     
        
Mimicking Marc-o's file directroy structure:    
baseDir <- sharedPathAn     
qiime2 <- paste(sharedPathAn, "qiime2", sep = "/")     
trimmed <- paste(sharedPathAn, "trimmed", sep = "/")     
logs <- paste(sharedPathAn, "logs", sep = "/")     
```{r}
# Make a directory for the trimmed data
trimmedData <- paste(sharedPathAn, "trimmed", sep = "/")
dir.create(trimmedData, showWarnings = TRUE, recursive = FALSE)

# Make a directory to hold the log files generated by itsxpress:
itsxpressLogs <- paste(sharedPathAn, "logs/itsxpress", sep = "/")
dir.create(itsxpressLogs, showWarnings = TRUE, recursive = TRUE)

qiime2Dir <- paste(sharedPathAn, "qiime2", sep = "/")
dir.create(qiime2Dir, showWarnings = TRUE, recursive = FALSE)
```

### Step 2:       
Retrieve ITS1 part of the amplicons using ITSxpress (includes trimming regions and export)      
Run ITSxpress on the raw fastq reads:   
     
**Note:** For the ITS2 region, itsxpress does not recognise the ITS2 amplicon regions and nothing is returned - no OTU table for ITS2 can be generated. Marco is looking into this issue to see if there is a sequence pattern at the ends that is inhibiting the correct processing of the sequences by itsxpress?? Possible overfitlering?? Direct checing of the sequences shows us that the ITS2 regions are in fact present, and there are many sequences observed. Problem appears to be random, and has worked with some sets of data but not others.... possible barcode issue??    
```{r}
prefix <- "ITSxpress_ITSF"
cmd <- paste("conda activate qiime2-2020.8 && itsxpress ",  
             " --fastq ", metadataITS$metadataITSF_files, 
             " --single_end ",
             " --outfile ", paste(trimmedData, "/ITSF_trimmed.", metadataITS$barcode, ".fastq", sep = ""),
             " --region ITS1 --taxa Fungi --cluster_id 0.995 ",
             " --log ", paste(itsxpressLogs, "/ITSF_trimmed.", metadataITS$barcode, ".log", sep = ""),
             " && conda deactivate ",
             sep = "")  
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
**To remove the output files after you are done:**
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

Running itsxpress on the itsf b56 reads separately, since they do not have a pair with the reverse reads for this barcode and are therefore not in the final metadataITS table, maybe they should be there?
```{r}
cmd <- paste("itsxpress ",  
             " --fastq ", metadataITSF$metadataITSF_files[56], 
             " --single_end ",
             " --outfile ", paste(trimmedData, "/ITSF_trimmed.", metadataITSF$barcode[56], ".fastq", sep = ""),
             " --region ITS1 --taxa Fungi --cluster_id 0.995 ",
             " --log ", paste(itsxpressLogs, "/ITSF_trimmed.", metadataITSF$barcode[56], ".log", sep = ""),
             sep = "")  
# Run the above command directly on the command line, since it's just for one file and shouldn't take long.
```

Add the path to the trimmed fastq files and a column to set unique sample names based on the filename/sequencing run and sample barcode number:
```{r}
metadataITSF$trimmedPath <- paste(trimmedData, "/ITSF_trimmed.", metadataITSF$barcode, ".fastq", sep = "")
metadataITSF$SampleID <- paste("ITSF_OAK_2019Plate1", metadataITSF$barcode, sep = "_")
```

Create a manifest file that qiime2 will use to import our fastq data and write it to a tsv file:
```{r}
library("data.table")
manifest <- metadataITSF[, .('sample-id' = SampleID, 'absolute-filepath' = trimmedPath)]

write.table(manifest, file = paste(sharedPathAn, "qiime2_import_manifest.tsv", sep = "/"), 
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)
```

# import fastq files
qiime tools import \
    --type 'SampleData[SequencesWithQuality]' \
    --input-path "${baseDir}"/fastq \
    --output-path "${baseDir}"/qiime2/demux-single-end.qza \
    --input-format CasavaOneEightSingleLanePerSampleDirFmt
```{r}
prefix <- "qiimeImport"
cmd <- paste("conda activate qiime2-2020.8 && ",
             "qiime tools import ",
             " --type 'SampleData[SequencesWithQuality]' ",
             " --input-path ", paste(sharedPathAn, "qiime2_import_manifest.tsv", sep = "/"),
             " --output-path ", paste(qiime2Dir, "/demux-single-end.qza", sep = ""),
             " --input-format SingleEndFastqManifestPhred33V2 ",
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
Create a file to visualise the qiime2 fastq files imported:
```{r}
prefix <- "qiimeVisualiseImport"
cmd <- paste("conda activate qiime2-2020.8 && ",
             " qiime demux summarize ",
             " --i-data  ", paste(qiime2Dir, "/demux-single-end.qza", sep = ""),
             " --o-visualization ", paste(qiime2Dir, "/demux-single-end.qzv", sep = ""),
             " --verbose ", 
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
To view demux-single-end.qzv, open https://view.qiime2.org/ with your browser and drag the file into the window at the top of the page.     
    
**To remove the output files after you are done:**
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

Denoise the sequences with dada2 within qiime2:    
- corrects sequencing errors    
- removes chimeras    
- clusters sequences at 100% similarity    
- outputs an asv table and the representative sequences
```{r}
prefix <- "qiimeDADA2deNoiseSingle"
cmd <- paste("conda activate qiime2-2020.8 && ",
             " qiime dada2 denoise-single ",
             " --i-demultiplexed-seqs ", paste(qiime2Dir, "/demux-single-end.qza", sep = ""),
             " --p-trim-left 0 ",
             " --p-trunc-len 0 ",
             " --o-representative-sequences ", paste(qiime2Dir, "/rep-seqs-dada2.qza", sep = ""),
             " --o-table ", paste(qiime2Dir, "/table-dada2.qza", sep = ""),
             " --o-denoising-stats ", paste(qiime2Dir, "/stats-dada2.qza", sep = ""),
             " --p-n-threads 20 ", 
             " --verbose ", 
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
Export the dada2 results:    
```{r}
# Export ASV (OTU-like table) table
prefix <- "qiimeExport"
cmd <- paste("conda activate qiime2-2020.8 && ",
             " mkdir ", paste(sharedPathAn, "phyloseq", sep = "/"),
             " && qiime tools export ",
             " --input-path ", paste(qiime2Dir, "/table-dada2.qza", sep = ""),
             " --output-path ", paste(sharedPathAn, "phyloseq", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

```{r}
# Convert biom format to tsv format
prefix <- "qiimeBiomConvert"
cmd <- paste("conda activate qiime2-2020.8 && ",
             " biom convert ",
             " -i ", paste(sharedPathAn, "phyloseq", "feature-table.biom", sep = "/"), 
             " -o ", paste(sharedPathAn, "phyloseq", "otu_table.tsv", sep = "/"),
             " --to-tsv && cd ", paste(sharedPathAn, "phyloseq", sep = "/"),
             " && sed -i '1d' otu_table.tsv && sed -i 's/#OTU ID//' otu_table.tsv && cd .. ",
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```

```{r}
# Export representative sequences
prefix <- "qiimeRepSeqsExport"
cmd <- paste("conda activate qiime2-2020.8 && ",
             " qiime tools export ",
             " --input-path ", paste(qiime2Dir, "/rep-seqs-dada2.qza", sep = ""),
             " --output-path ", paste(sharedPathAn, "phyloseq", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
```{r}
RemoveQsubTempFiles(sharedPathAn, prefix)
```


Set up the qiime2 UNITE database using UNITE 2018-2019 that Marc-o used:
```{r}
uniteDBDir2019 <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/Databases/UNITE_2018-11-18_qiimeReleaseDB"
```

```{r}
# Import the UNITE reference sequences into QIIME2.
prefix <- "qiimeUniteImport"
cmd <- paste("conda activate qiime2-2020.8 && ",
             " qiime tools import ",
             " --type FeatureData[Sequence] ",
             " --input-path ", paste(uniteDBDir2019, "sh_refs_qiime_ver8_99_02.02.2019.fasta", sep = "/"),
             " --output-path ", paste(sharedPathAn, "phyloseq", "unite-ver8_99_02.02.2019.qza", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```


```{r}
# Import the taxonomy file.
prefix <- "qiimeTaxUniteImport"
cmd <- paste("conda activate qiime2-2020.8 && ",
             " qiime tools import ",
             " --type FeatureData[Taxonomy] ",
             " --input-path ", paste(uniteDBDir2019, "sh_taxonomy_qiime_ver8_99_02.02.2019.txt", sep = "/"),
             " --output-path ", paste(sharedPathAn, "phyloseq", "unite-ver8-taxonomy_99_02.02.2019.qza", sep = "/"),
             " --input-format HeaderlessTSVTaxonomyFormat ",
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

```{r}
# Train the classifier
prefix <- "qiimeTrainUnite"
cmd <- paste("conda activate qiime2-2020.8 && ",
             " qiime feature-classifier fit-classifier-naive-bayes ",
             " --i-reference-reads ", paste(sharedPathAn, "phyloseq", "unite-ver8_99_02.02.2019.qza", sep = "/"), 
             " --i-reference-taxonomy ", paste(sharedPathAn, "phyloseq", "unite-ver8-taxonomy_99_02.02.2019.qza", sep = "/"),
             " --o-classifier ", paste(sharedPathAn, "phyloseq", "unite-ver8-classifier_99_02.02.2019.qza", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

```{r}
prefix <- "qiimeClassifyFeature"
cmd <- paste("conda activate qiime2-2020.8 && ",
             " qiime feature-classifier classify-sklearn ",
             " --i-classifier ", paste(sharedPathAn, "phyloseq", "unite-ver8-classifier_99_02.02.2019.qza", sep = "/"),
             " --i-reads ", paste(qiime2Dir, "rep-seqs-dada2.qza", sep = "/"),
             " --o-classification ", paste(qiime2Dir, "taxonomy-single-end.qza", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```
```{r}
prefix <- "qiimeFeatureExport"
cmd <- paste("conda activate qiime2-2020.8 && ",
             " qiime tools export ",
             " --input-path ", paste(qiime2Dir, "taxonomy-single-end.qza", sep = "/"),
             " --output-path ", paste(sharedPathAn, "phyloseq", sep = "/"),
             " && conda deactivate ", sep = "")
suffix <- ".sub"; cat(bashDirections); MakeQsubs(cmd, prefix, suffix)
```

To get the ASV/OTU tablewith taxonomy column at the end, do we combine the otu_table.tsv and taxonomy.tsv together by feature ID row?
```{r}
list.files(path = paste(sharedPathAn, "phyloseq", sep = "/"))
library(data.table)
library(phyloseq)
otuTbl <- fread(paste(sharedPathAn, "phyloseq", "otu_table.tsv", sep = "/"))
taxTbl <- fread(paste(sharedPathAn, "phyloseq", "taxonomy.tsv", sep = "/"))


# Open the taxonomy and change the header. When you open it, you’ll see the header looks like this: 
# Feature ID	Taxon	Confidence
# where the spaces are tabs. You need to change it to this:
# otu-id	taxonomy	Confidence

setnames(taxTbl, "Feature ID", "otu-id")
setnames(taxTbl, "Taxon", "taxonomy")

setnames(otuTbl, "V1", "otu-id")

setkey(otuTbl, "otu-id")
setkey(taxTbl, "otu-id")

otuTaxTbl <- merge(otuTbl, taxTbl)
otuTaxTbl$Confidence <- NULL

write.table(otuTaxTbl, file = paste(sharedPathAn, "otuTax_table.tsv", sep = "/"), 
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)
```
Save image:
```{r}
save.image(paste(imageDirPath, baseImage, sep = ""))
```


Create a directory to store RAM and other analyses:
```{r}
metaAnalysesPath <- paste(sharedPathAn, "metaAnalyses", sep = "/")
dir.create(metaAnalysesPath, showWarnings = TRUE, recursive = FALSE)
```


```{r}
library("RAM")
taxFillPath <- paste(metaAnalysesPath, "taxFill", sep = "/")
dir.create(taxFillPath, showWarnings = TRUE, recursive = FALSE)

temp <- read.table(paste(sharedPathAn, "otuTax_table.tsv", sep = "/"), sep = "\t", header = TRUE, dec = ".", 
                   comment.char = "", quote = "", stringsAsFactors = TRUE,
                   as.is = TRUE, colClasses=c("taxonomy"="character"))

row.names(temp) <- temp$otu.id
temp$otu.id <- NULL

temp <- tax.fill(temp, downstream=TRUE)

write.table(temp, file=paste(taxFillPath, "ITS1F.table.taxfill.tsv", sep = "/"),
            append = FALSE, sep = "\t", row.names = FALSE, quote = FALSE)
```


Create diversity directory and path. List your datasets for diversity indexes. Adds a bunch of columns in a new diversity metadata file.
Indices obtained (in this specific order) are: Spec Number,  Simpson data,  Inv simpson data,	Shannon data,	Simpson eveness,	Shannon eveness,	Simpson true diversity,	shannon true diversity,	chao,	ACE.
```{r}
diversityPath <- paste(metaAnalysesPath, "diversity", sep = "/")
dir.create(diversityPath, showWarnings = TRUE, recursive = FALSE)

metaTemp <- metadataITSF
rownames(metaTemp) <- colnames(temp)[-ncol(temp)] #seems to work
temp2 <- OTU.diversity(list(data=temp), metaTemp)
write.table(temp2, file=paste(diversityPath, "OWPlate1_ITS1F.meta.div.tsv", sep = "/"),
            append = FALSE, sep = "\t", row.names = FALSE, quote=FALSE)
```


```{r}
diversityPlotPath <- paste(metaAnalysesPath, "diversityPlot", sep = "/")
dir.create(diversityPlotPath, showWarnings = TRUE, recursive = FALSE)


```


Rarefy your tables:
```{r}
dir.create(paste(pathDataAnMC, "rarefaction/", sep = ""),
           showWarnings = TRUE,
           recursive    = FALSE)
RarefactionPathMC <- paste(pathDataAnMC, "rarefaction/", sep = "")
##############
#            #
#    ITS     #
#            #
############## works
temp <-  read.table(paste(pathDataAnMC, "DP_merged_ITS.table.tsv", sep = ""),
                      sep = "\t", header = TRUE, dec = ".", comment.char = "", quote = "", stringsAsFactors = TRUE,
                      as.is = TRUE, check.names = FALSE, colClasses=c("row.names"="character", "taxonomy"="character"))
  temp$row.names <- NULL 
rrf <- OTU.rarefy(list(data=temp), sample=NULL)
# Write out the dataframe:
  write.table(rrf, file=paste(RarefactionPathMC, "ITS.rarefy.2.tsv", sep = ""),
                               append    = FALSE,
                               sep       = "\t",
                               row.names = FALSE,
                               quote=FALSE)
  
temp2 <-  read.table(paste(pathDataAnMC, "DP_merged_ITS.table.tsv", sep = ""),
                      sep = "\t", header = TRUE, dec = ".", comment.char = "", quote = "", stringsAsFactors = TRUE,
                      as.is = TRUE, check.names = FALSE, colClasses=c("taxonomy"="character"))
temp2$row.names <- NULL
temp2$taxonomy <- NULL
 
# Open file-handle to get ready to make the png graph:
  png(filename=paste(RarefactionPathMC, "ITS", ".RAREFplot.2.png", sep = ""),
      width = 1200, height = 400, units = "px")
rrfPLOT <- rarecurve(temp2, step =10, xlab = "Sequence nb per sample", ylab = "Species OTUs") 
# Close the png graph file handle:
  dev.off()
```

##TaxAbund
### group.abund.Taxa
# 1. generate the list of taxa desired with core taxa, then generate the plots with group.abund.Taxa
```{r}   
library("RAM")
# make new directory:
dir.create(paste(pathDataAnMC, "groupTaxaBarPlot/", sep = ""),
           showWarnings = TRUE,
           recursive    = FALSE)
taxaBarPlotPathMC <- paste(pathDataAnMC, "groupTaxaBarPlot/", sep = "")
##############
#            #
#    ITS     #
#            #
##############
tabletemp <-  read.table(paste(taxFillPathMC, "ITS.table.taxFill.2.tsv", sep = ""),
                      sep = "\t", header = TRUE, dec = ".", comment.char = "", quote = "", stringsAsFactors = TRUE,
                      as.is = TRUE, check.names = FALSE, colClasses=c("row.names"="character","taxonomy"="character"))
  
tabletemp$row.names <- NULL 
  
metatemp <-  read.table(paste(pathDataAnMC, "DP_merged_ITS.meta.tsv", sep = ""),
                      sep = "\t", header = TRUE, dec = ".", comment.char = "", quote = "", stringsAsFactors = TRUE,
                      as.is = TRUE, check.names = FALSE)
metatemp$row.names <- NULL 
rownames(metatemp) <- colnames(tabletemp)[-ncol(tabletemp)] #seems to work  
#species_all
coreTaxaITSdate <- core.Taxa(data = list(data = tabletemp), is.OTU = TRUE, meta = metatemp, rank = "s", drop.unclassified = FALSE, meta.factor = "CollectionDate", percent = 0)
#taxaDate <- coreTaxaITSdate$data$CollectionDate$taxa
# Open file-handle to get ready to make the png graph:
  png(filename=paste(taxaBarPlotPathMC, "ITS", ".species.all.2.png", sep = ""),
      width = 10000, height = 10000, units = "px", res = 100)
myGroup.abund.Taxa(data = list(data = tabletemp),
                    is.OTU = TRUE,
                    rank = "s",    
                    drop.unclassified = FALSE,
                    meta = metatemp,
                    meta.factor = "CollectionDate",
                    taxa = coreTaxaITSdate,
                    main = "Species_all")
                
# Close the png graph file handle:
  dev.off()
#genera_all
coreTaxaITSdate <- core.Taxa(data = list(data = tabletemp), is.OTU = TRUE, meta = metatemp, rank = "g", drop.unclassified = FALSE, meta.factor = "CollectionDate", percent = 0)
#taxaDate <- coreTaxaITSdate$data$CollectionDate$taxa
# Open file-handle to get ready to make the png graph:
  png(filename=paste(taxaBarPlotPathMC, "ITS", ".genera.all.bestres.png", sep = ""),
      width = 8000, height = 8000, units = "px", res = 300)
MyGroup.abund.taxa.version2(data = list(data = tabletemp),
                    is.OTU = TRUE,
                    rank = "g",    
                    drop.unclassified = FALSE,
                    meta = metatemp,
                    meta.factor = "CollectionDate",
                    taxa = coreTaxaITSdate,
                    main = "Genera_all")
                
# Close the png graph file handle:
  dev.off() 
  
  
#family_all
coreTaxaITSdate <- core.Taxa(data = list(data = tabletemp), is.OTU = TRUE, meta = metatemp, rank = "f", drop.unclassified = FALSE, meta.factor = "CollectionDate", percent = 0)
#taxaDate <- coreTaxaITSdate$data$CollectionDate$taxa
# Open file-handle to get ready to make the png graph:
  png(filename=paste(taxaBarPlotPathMC, "ITS", ".family.all.2.png", sep = ""),
      width = 3000, height = 2000, units = "px", res = 100)
myGroup.abund.Taxa(data = list(data = tabletemp),
                    is.OTU = TRUE,
                    rank = "f",    
                    drop.unclassified = FALSE,
                    meta = metatemp,
                    meta.factor = "CollectionDate",
                    taxa = coreTaxaITSdate,
                    main = "Family_all")
                
# Close the png graph file handle:
  dev.off()   
  
 #order_all
coreTaxaITSdate <- core.Taxa(data = list(data = tabletemp), is.OTU = TRUE, meta = metatemp, rank = "o", drop.unclassified = FALSE, meta.factor = "CollectionDate", percent = 0)
#taxaDate <- coreTaxaITSdate$data$CollectionDate$taxa
# Open file-handle to get ready to make the png graph:
  png(filename=paste(taxaBarPlotPathMC, "ITS", ".order.all.2.png", sep = ""),
      width = 3000, height = 2000, units = "px", res = 100)
myGroup.abund.Taxa(data = list(data = tabletemp),
                    is.OTU = TRUE,
                    rank = "o",    
                    drop.unclassified = FALSE,
                    meta = metatemp,
                    meta.factor = "CollectionDate",
                    taxa = coreTaxaITSdate,
                    main = "Order_all")
                
# Close the png graph file handle:
  dev.off()   
   
  
 #class_all
coreTaxaITSdate <- core.Taxa(data = list(data = tabletemp), is.OTU = TRUE, meta = metatemp, rank = "c", drop.unclassified = FALSE, meta.factor = "CollectionDate", percent = 0)
#taxaDate <- coreTaxaITSdate$data$CollectionDate$taxa
# Open file-handle to get ready to make the png graph:
  png(filename=paste(taxaBarPlotPathMC, "ITS", ".class.all.2.png", sep = ""),
      width = 3000, height = 2000, units = "px", res = 100)
myGroup.abund.Taxa(data = list(data = tabletemp),
                    is.OTU = TRUE,
                    rank = "c",    
                    drop.unclassified = FALSE,
                    meta = metatemp,
                    meta.factor = "CollectionDate",
                    taxa = coreTaxaITSdate,
                    main = "Class_all")
                
# Close the png graph file handle:
  dev.off()     
  
 #phyla_all
coreTaxaITSdate <- core.Taxa(data = list(data = tabletemp), is.OTU = TRUE, meta = metatemp, rank = "p", drop.unclassified = FALSE, meta.factor = "CollectionDate", percent = 0, number=5)
#taxaDate <- coreTaxaITSdate$data$CollectionDate$taxa
# Open file-handle to get ready to make the png graph:
  png(filename=paste(taxaBarPlotPathMC, "ITS", ".phyla.all.2.png", sep = ""),
      width = 3000, height = 2000, units = "px", res = 100)
myGroup.abund.Taxa(data = list(data = tabletemp),
                    is.OTU = TRUE,
                    rank = "p",    
                    drop.unclassified = FALSE,
                    meta = metatemp,
                    meta.factor = "CollectionDate",
                    taxa = coreTaxaITSdate,
                    main = "phyla_all")
                
# Close the png graph file handle:
  dev.off()  

# GroupTaxabar
dir.create(paste(pathDataAnMC, "groupAbundnbBarPlot/", sep = ""),
           showWarnings = TRUE,
           recursive    = FALSE)
abundNbBarPlotPathMC <- paste(pathDataAnMC, "groupAbundnbBarPlot/", sep = "")
##############
#            #
#    ITS     #
#            #
############## works
tabletemp <-  read.table(paste(taxFillPathMC, "ITS.table.taxFill.2.tsv", sep = ""),
                      sep = "\t", header = TRUE, dec = ".", comment.char = "", quote = "", stringsAsFactors = TRUE,
                      as.is = TRUE, check.names = FALSE, colClasses=c("taxonomy"="character"))
tabletemp$row.names <- NULL 
 
metatemp <-  read.table(paste(pathDataAnMC, "DP_merged_ITS.meta.tsv", sep = ""),
                      sep = "\t", header = TRUE, dec = ".", comment.char = "", quote = "", stringsAsFactors = TRUE,
                      as.is = TRUE)
metatemp$row.names <- NULL
rownames(metatemp) <- colnames(tabletemp)[-ncol(tabletemp)] #seems to work
#species
# Open file-handle to get ready to make the plot:
  png(filename=paste(abundNbBarPlotPathMC, "ITS", ".gr.abund.species.top30.2.png", sep = ""),
      width = 3000, height = 3000, units = "px", pointsize = 12, res = 300)
group.abundance.meta(data=(list(data=tabletemp)), rank="s", top = 30, count =  TRUE, drop.unclassified = FALSE, cex.x = 10, main = "Top 30 counts of taxonomic groups at the species level", meta = metatemp, meta.factor = c("CollectionDate"))
# Close the png graph file handle:
  dev.off()
  #genus
  # Open file-handle to get ready to make the plot:
  png(filename=paste(abundNbBarPlotPathMC, "ITS", ".gr.abund.genera.top15.2.png", sep = ""),
      width = 3000, height = 3000, units = "px", pointsize = 12, res = 300)
group.abundance.meta(data=(list(data=tabletemp)), rank="g", top = 15, count =  TRUE, drop.unclassified = FALSE, cex.x = 10, main = "Top 15 counts of taxonomic groups at the genus level ", meta = metatemp, meta.factor = c("CollectionDate"))
# Close the png graph file handle:
  dev.off()
  #family
  # Open file-handle to get ready to make the plot:
  png(filename=paste(abundNbBarPlotPathMC, "ITS", ".gr.abund.family.top15.2.png", sep = ""),
      width = 3000, height = 3000, units = "px", pointsize = 12, res = 300)
group.abundance.meta(data=(list(data=tabletemp)), rank="f", top = 15, count =  TRUE, drop.unclassified = FALSE, cex.x = 10, main = "Top 15 counts of taxonomic groups at the family level", meta = metatemp, meta.factor = c("CollectionDate"))
# Close the png graph file handle:
  dev.off()
   
   #order
  # Open file-handle to get ready to make the plot:
  png(filename=paste(abundNbBarPlotPathMC, "ITS", ".gr.abund.order.top10.2.png", sep = ""),
      width = 3000, height = 3000, units = "px", pointsize = 12, res = 300)
group.abundance.meta(data=(list(data=tabletemp)), rank="o", top = 10, count =  TRUE, drop.unclassified = FALSE, cex.x = 10, main = "Top 10 counts of taxonomic groups at the order level", meta = metatemp, meta.factor = c("CollectionDate"))
# Close the png graph file handle:
  dev.off()
  
  
 #class
  # Open file-handle to get ready to make the plot:
  png(filename=paste(abundNbBarPlotPathMC, "ITS", ".gr.abund.class.top10.2.png", sep = ""),
      width = 3000, height = 3000, units = "px", pointsize = 12, res = 300)
group.abundance.meta(data=(list(data=tabletemp)), rank="c", top = 10, count =  TRUE, drop.unclassified = FALSE, cex.x = 10, main = "Top 10 counts of taxonomic groups at the class level", meta = metatemp, meta.factor = c("CollectionDate"))
# Close the png graph file handle:
  dev.off()
  
  #phyla
  # Open file-handle to get ready to make the plot:
  png(filename=paste(abundNbBarPlotPathMC, "ITS", ".gr.abund.phyla.top5.2.png", sep = ""),
      width = 3000, height = 3000, units = "px", pointsize = 12, res = 300)
group.abundance.meta(data=(list(data=tabletemp)), rank="p", top = 5, count =  TRUE, drop.unclassified = FALSE, cex.x = 10, main = "Top 50 OTU abundance (count) at the phyla level ", meta = metatemp, meta.factor = c("CollectionDate"))
# Close the png graph file handle:
  dev.off()
      
```

species query with metaresultextractor
 
 
```{r}  
---
  title: "species.query.marie-claude"
author: "E.T."
date: "5/6/2018"
#######################
#########################
#                       #
#    ITS by species     #
#                       #
#########################
#Define your 6 arguments to run the perl script
query_in <- paste(WordMatchDir, "genusQuery.txt", sep = "")
otu_in <- paste(pathDataAnMC, "DP_merged_ITS.table.tsv", sep = "")
meta_in <- paste(pathDataAnMC, "DP_merged_ITS.meta.tsv", sep = "")
fasta_in <- paste(pathDataAnMC, "ref.seqs.ITS.DP.fasta", sep = "")
table_taxon_out <- paste(WordMatchDir, "ITS_species_query_by_taxon.MC.pollen.tsv", sep = "")
table_sample_out <- paste(WordMatchDir, "ITS_species_query_by_sample.MC.pollen.tsv", sep = "")
fasta_out_folder <- paste(WordMatchDir, "/FastaOut/ITS_species_query_MC_pollen", sep = "")
cmd <- paste("perl", paste(ScriptsPath, "metaResultExtractorGOOD.pl", sep = ""),
             query_in,
             otu_in,
             meta_in,
             fasta_in,
             table_taxon_out,
             table_sample_out,
             fasta_out_folder,
             sep = " ")
system(cmd)
#then, use the cluster script
#########################
#                       #
#    ITS by genuses     #
#                       #
#########################
#Define your 6 arguments to run the perl script
query_in <- paste(WordMatchDir, "genusQuery.txt", sep = "")
otu_in <- paste(pathDataAn, "ITS.table.tsv", sep = "")
meta_in <- paste(pathDataAn, "ITS.meta.tsv", sep = "")
fasta_in <- paste(pathDataAn, "new_refseqs_ITS.fna", sep = "")
table_taxon_out <- paste(WordMatchDir, "ITS_genus_query_by_taxon2.tsv", sep = "")
table_sample_out <- paste(WordMatchDir, "ITS_genus_query_by_sample2.tsv", sep = "")
fasta_out_folder <- paste(WordMatchDir, "/FastaOut/ITS_genus_query", sep = "")
cmd <- paste("perl", paste(ScriptsPath, "metaResultExtractorGOOD.pl", sep = ""),
             query_in,
             otu_in,
             meta_in,
             fasta_in,
             table_taxon_out,
             table_sample_out,
             fasta_out_folder,
             sep = " ")
system(cmd)
####
#more species/genera to the list
#Define your 6 arguments to run the perl script
query_in <- paste(WordMatchDir, "genus.and.species.quesry.pollen.txt", sep = "")
otu_in <- paste(pathDataAnMC, "DP_merged_ITS.table.tsv", sep = "")
meta_in <- paste(pathDataAnMC, "DP_merged_ITS.meta.tsv", sep = "")
fasta_in <- paste(pathDataAnMC, "ref.seqs.ITS.DP.fasta", sep = "")
table_taxon_out <- paste(WordMatchDir, "ITS_species_ag.query_by_taxon.MC.pollen.tsv", sep = "")
table_sample_out <- paste(WordMatchDir, "ITS_species_ag.query_by_sample.MC.pollen.tsv", sep = "")
fasta_out_folder <- paste(WordMatchDir, "/FastaOut/ITS_species_query_MC_pollen", sep = "")
cmd <- paste("perl", paste(ScriptsPath, "metaResultExtractorGOOD.pl", sep = ""),
             query_in,
             otu_in,
             meta_in,
             fasta_in,
             table_taxon_out,
             table_sample_out,
             fasta_out_folder,
             sep = " ")
system(cmd)
#oomycetes
#more species/genera to the list
#Define your 6 arguments to run the perl script
query_in <- paste(WordMatchDir, "oomycete.query.pollen.txt", sep = "")
otu_in <- paste(pathDataAnMC, "DP_merged_OM.table.tsv", sep = "")
meta_in <- paste(pathDataAnMC, "DP_merged_OM.meta.tsv", sep = "")
fasta_in <- paste(pathDataAnMC, "OM.ref.seqs.DP.fasta", sep = "")
table_taxon_out <- paste(WordMatchDir, "omycete_query_by_taxon.MC.pollen.tsv", sep = "")
table_sample_out <- paste(WordMatchDir, "oomycetes_query_by_sample.MC.pollen.tsv", sep = "")
fasta_out_folder <- paste(WordMatchDir, "/FastaOut/ITS_species_query_MC_pollen", sep = "")
cmd <- paste("perl", paste(ScriptsPath, "metaResultExtractorGOOD.pl", sep = ""),
             query_in,
             otu_in,
             meta_in,
             fasta_in,
             table_taxon_out,
             table_sample_out,
             fasta_out_folder,
             sep = " ")
system(cmd)
###oom curated
#oomycetes
#more species/genera to the list
#Define your 6 arguments to run the perl script
query_in <- paste(WordMatchDir, "oomycete.query.pollen.txt", sep = "")
otu_in <- paste(pathDataAnMC, "DP_merged_OM.curated.tsv", sep = "")
meta_in <- paste(pathDataAnMC, "DP_merged_OM.meta.tsv", sep = "")
fasta_in <- paste(pathDataAnMC, "OM.ref.seqs.DP.cur.fasta", sep = "")
table_taxon_out <- paste(WordMatchDir, "omycete_query_by_taxon.MC.pollen.cur.tsv", sep = "")
table_sample_out <- paste(WordMatchDir, "oomycetes_query_by_sample.MC.pollen.cur.tsv", sep = "")
fasta_out_folder <- paste(WordMatchDir, "/FastaOut/ITS_species_query_MC_pollen", sep = "")
cmd <- paste("perl", paste(ScriptsPath, "metaResultExtractorGOOD.pl", sep = ""),
             query_in,
             otu_in,
             meta_in,
             fasta_in,
             table_taxon_out,
             table_sample_out,
             fasta_out_folder,
             sep = " ")
system(cmd)
#plants
#oomycetes
#more species/genera to the list
#Define your 6 arguments to run the perl script
query_in <- paste(WordMatchDir, "plants.query.pollen.txt", sep = "")
otu_in <- paste(pathDataAnMC, "DP_merged_plants.table.tsv", sep = "")
meta_in <- paste(pathDataAnMC, "DP_merged_plants.meta.tsv", sep = "")
fasta_in <- paste(pathDataAnMC, "Plants.ref.seqs.DP.fasta", sep = "")
table_taxon_out <- paste(WordMatchDir, "plants_query_by_taxon.MC.pollen.tsv", sep = "")
table_sample_out <- paste(WordMatchDir, "plants_query_by_sample.MC.pollen.tsv", sep = "")
fasta_out_folder <- paste(WordMatchDir, "/FastaOut/plants_species_query_MC_pollen", sep = "")
cmd <- paste("perl", paste(ScriptsPath, "metaResultExtractorGOOD.pl", sep = ""),
             query_in,
             otu_in,
             meta_in,
             fasta_in,
             table_taxon_out,
             table_sample_out,
             fasta_out_folder,
             sep = " ")
system(cmd)